Feat/delivery order join

ohk8 opened the pull request:
#41 

#44 

order와 delivery join컬럼 및 관련 코드 수정했습니다. (+ #43 리뷰 반영 살짝 수정된 부분도 포함되었습니다:D )

일단 테스트는 통과해서 올리는데 코드 오류가 있을지... 확인 부탁드립니다.

jongeuni committed:
[ADD #38] MenuCreateRequestDto

jongeuni committed:
[ADD #38] menuCreate method to MenuController

jongeuni committed:
[ADD #38] CreateMenu method to MenuService and set isHidden default false

jongeuni committed:
[ADD #38] MenuProduct save logic

jongeuni committed:
[ADD #38] @Valid to MenuController

jongeuni committed:
[ADD #38] @RequestBody to menuCreate method

jongeuni committed:
[MODIFY #38] MenuCreateRequestDto products type change List -> Map

메뉴에 속하는 상품이 여러개일 수 있기 때문에 상품 갯수도 함께 받도록 변경

jongeuni committed:
[MODIFY #38] Use Elvis operation instead of orElseThrow

ohjiyeon123 committed:
exception관련클래스 분리, post->put 주소수정(restapi)

ohjiyeon123 committed:
주문상태변경 관련 테스트코드 추가

ohjiyeon123 committed:
test-exception코드 추가

ohjiyeon123 committed:
test-exception코드 추가

ohjiyeon123 committed:
Merge branch 'feat/delivery' of https://github.com/ohjiyeon3275/kitchen-force into feat/delivery

ohjiyeon123 committed:
#41 #44 delivery-order join 추가 및 관련 코드 수정

ohjiyeon123 committed:
delivery status변경시 불필요한 riderId코드 삭제

ohjiyeon123 committed:
riderId관련 test코드 수정

humanitas03 reviewed (commented):
confilict가 나는 부분이 있네요..

humanitas03 commented on src/main/kotlin/com/kitchenforce/domain/delivery/DeliveryAddress.kt:
```diff
@@ -20,12 +15,16 @@ class DeliveryAddress(
 
     @Column(nullable = false)
-    var status: String,
+    var accountStatus: String,
```
not null 필드가 아니면서 테스트에서 빈 문자열을 넣고, 로직에서 따로 이걸 쓰는 부분이 없는거 같은데 이건 어떤 의미의 컬럼인가요?

ohk8 replied:
이부뷴 이전 회의에서 나왔던 내용 이기도 합니당!-! (DeliveryAddress -> Delivery)
어떤 테스트에서는 중요치 않다고 생각해 빈 스트링으로 주었던 것 같습니다... 😂 

DeliveryAddress의 status 기능을 1) 계정상태 2) 주문상태 로 나눠서 관리해야겠다고 생각습니다.

1)의 기능은 해당 주소의 고객의 정보 ex) 전화번호등 가 바뀐다면 이전 고객의 정보를 기록차원에서 남겨두고 상태만 휴면(?)으로 변경할까 했습니다. (accountStatus가 active인 account만 검색)

계정이 휴면일때의 exception은 다음 기능추가때 만들어보겠습니다 😆

ohk8 commented:
> confilict가 나는 부분이 있네요..

Order 테이블과 조인시에 Order와 Orderdto수정한 부분이 있어서 그럴 것 같습니다. 충분한 코드 추가 검토후 머지후에 conflict나는 부분은 수정해서 다시 올리겠습니다

seongkyu-lim committed:
add: @valid for numOfGuests

방문한 손님 수는 0 이상이어야 한다.

seongkyu-lim committed:
add: getEmptyness logic

해당 테이블이 빈 테이블인지, 이용 중인 테이블인지 체크.
<요구사항>
빈 테이블을 해지할 수 있다.
빈 테이블로 설정할 수 있다.
방문한 손님 수를 변경할 수 있다.
방문한 손님 수가 올바르지 않으면 변경할 수 없다.
방문한 손님 수는 0 이상이어야 한다.
빈 테이블은 방문한 손님 수를 변경할 수 없다.

seongkyu-lim committed:
refactor: emptyness -> emptiness

seongkyu-lim committed:
add: updateStatus api

<요구사항>

주문을 접수한다.
접수 대기 중인 주문만 접수할 수 있다.
주문을 서빙한다.
접수된 주문만 서빙할 수 있다.
주문을 완료한다.
포장 및 매장 주문의 경우 서빙된 주문만 완료할 수 있다.
주문 테이블의 매장 주문이 완료되면 빈 테이블로 설정한다.
매장 주문이 완료되지 않았다면 빈 테이블로 설정하지 않는다.

jongeuni committed:
[MODIFY #38] Menu product request Change

jongeuni committed:
Merge pull request #39 from kotlin-serverside-study/feature/menu_create

Feature/menu create

seongkyu-lim committed:
refactor: order create api

<요구사항>
메뉴가 없으면 등록할 수 없다.
매장 주문을 제외한 주문의 경우 주문 항목의 수량은 0 이상이어야 한다.
숨겨진 메뉴는 주문할 수 없다.
주문한 메뉴의 가격은 실제 메뉴 가격과 일치해야 한다.

seongkyu-lim commented on src/main/kotlin/com/kitchenforce/service/DeliveryService.kt:
```diff
@@ -18,19 +18,19 @@ class DeliveryService(
         val riderOptional = riderRepository.findById(riderId)
```
findByIdorNull을 이용하는 것은 어떠신가요 ?

해서, rider?.let{} ?: 구문을 사용해보시는 것이 어떨가요 ? 

    val rider = riderRepository.findByIdOrNull(id)

    rider?.let{
        if(rider.deliveryAddress[deliveryAddressId.toInt()] == null){
            throw DeliveryException(DeliveryErrorCodeType.NOT_FOUND_DELIVERY)
        }

        val delivery = rider.deliveryAddress[deliveryAddressId.toInt()]

        if (delivery.status != "배달중") {
            throw DeliveryException(DeliveryErrorCodeType.NOT_MATCHED_DELIVERY_STATUS)
        }else{
            delivery.status = "배달완료"
            return deliveryAddressRepository.save(delivery)
    } ?:  throw DeliveryException(DeliveryErrorCodeType.NOT_FOUND_RIDER)

seongkyu-lim commented on src/main/kotlin/com/kitchenforce/service/DeliveryService.kt:
```diff
@@ -18,19 +18,19 @@ class DeliveryService(
 
         val rider = riderOptional.get()
 
         if(rider.deliveryAddress[deliveryAddressId.toInt()] == null){
```
코틀린의 elvis 연산자를 이용해보시는 것은 어떨까요 ??

seongkyu-lim commented:
주문 관련 로직 부분이 많이 변경되었는데, 변경하기 이전꺼에 코드를 짜주셨네요ㅠ .  pull하셔서 충돌 해결해주시고 짜주신 코드 다시 반영해야할 것 같습니다..!

seongkyu-lim commented on src/main/kotlin/com/kitchenforce/service/OrderTableService.kt:
```diff
@@ -34,13 +35,14 @@ class OrderTableService(
                 paymentMethod = orderDto.paymentMethod,
                 requirement = orderDto.requirement,
-                deliveryAddress = orderDto.deliveryAddress,
+                deliveryAddress = deliveryAddressId?.let { orderRepository.findByDeliveryAddressExists(it) },
```
deliveryAddress 엔티티를 생성하고 관리하는 것에 있어서 좀 더 고민이 필요할 것 같아요,

우선 deliveryAddress에서 배달관련된 전반적인 데이터를 담당하니 이름을 delivery로 변경하는 것이 어떨까요 ?

그리고 주문 생성 로직에서 배달 주문인 경우  delivery(deliveryAddress)를 생성해주는 것이 맞는 것 같은데 어떻게 생각하시나요 ?

해서 delivery의 상태 update로직은 배달 대기 -> 배달 진행 -> 배달 완료순으로 진행되며,  
"배달 대기에서 배달 진행으로의 상태 업데이트는 주문 상태가 주문 완료(Closed)상태일 때만 가능하다." 정도의 요구사항을 추가하여 update로직도 수정하는 것이 어떤가요 ?

delivery 도메인 부분이 기능 설계가 되어있지 않은 부분이라 수고가 많으십니다..!

seongkyu-lim commented on src/main/kotlin/com/kitchenforce/domain/orders/Order.kt:
```diff
@@ -27,8 +19,8 @@ class Order(
 
-    @Column
-    var deliveryAddress: String,
+    @OneToOne
```
충돌해결해보시면 아시겠지만, 한 집에서 여러개의 주문을 할 수 있다고 생각하여 @ManyToOne으로 관계설정을 하였습니다..!

humanitas03 replied:
컨플릭 해결하시고 머지 하시면 이 컨셉을 제가 올릴게요

seongkyu-lim committed:
add : test (orderService, orderTableService)

seongkyu-lim committed:
refactor : mapping order, orderTable

oneToOne 양방향 매핑에 대해 LAZY이슈가 존재하여 단방향 매핑으로 변경하고, order와 emptiness로 orderTable을 조회하는 방식으로 수정.

요구사항 : 주문 테이블의 모든 매장 주문이 완료되면 빈 테이블로 설정한다.

seongkyu-lim committed:
refactor: getEmptiness -> getTableInfo

seongkyu-lim committed:
refactor: updateStatus logic.

매장 식사인 경우 주문이 완료되면 emptiness -> true  save 로직 추가.

seongkyu-lim committed:
refactor:  using 'it' keyword

seongkyu-lim committed:
refactor:  patchmapping -> putmaaping / RequestParam -> Pathvariable

humanitas03 commented:
이 커밋은 너무 장수해서 기네스 북에 오르겟어요...

seongkyu-lim committed:
Merge pull request #46 from kotlin-serverside-study/feature/#44

Feature/order : issue #44

ohjiyeon123 committed:
Merge remote-tracking branch 'origin/feat/delivery-order-join'

# Conflicts:
#	src/main/kotlin/com/kitchenforce/domain/orders/Order.kt
#	src/main/kotlin/com/kitchenforce/domain/orders/dto/OrderDto.kt
#	src/main/kotlin/com/kitchenforce/service/OrderTableService.kt
#	src/test/kotlin/com/kitchenforce/domain/delivery/DeliveryAddressTest.kt
#	src/test/kotlin/com/kitchenforce/service/OrderTableServiceTest.kt

ohjiyeon123 committed:
conflict 수정
